### PR TITLE
feat: added custom validation for loginId/userId when tried to auto l…

### LINKED ISSUE #716: [FEATURE] FRONT-END: Validate loginId to be a valid uuid
### Description

Code: https://github.com/Dun-sin/Whisper/blob/1c1224036887cd9981d6c389e987569c07193976/client/src/context/AuthContext.jsx#L30

### Screenshots

_No response_

### Additional information

_No response_

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Gopal-001**
**My PR closes #716 **

# 👨‍💻 Changes proposed(What did you do ?)
- Added validateUserID for both client and server for validating the loginId.
- In client added check in AuthContext to prevent illegal usage or local storage auth.
- the validateUserID function in client checks weather
  - the id stored in local storage is for type email or anonymous login and validated the id based on respective patterns 
  - if type is email than we generated a hex id which is a (in our case) 24 char long string consisting of only [a-f0-9]
  - if type anonymous than id should be valid 12 char long string with characters of only [a-z0-9].
- in validateUserID server function added it as a module export which will verify the case 
  - where email and id is being given to add a user (i.e when an anonymous user tries to add profile),
  -  if only email (i.e id is undefined in case of email login add new user) no change required. 
  
Client function
```
export const validateUserID = (userID, userType) => {
  const userIDPattern = /^[a-z0-9]{12}$/;
  const userHexIdPattern = /^[a-f0-9]{24}$/;
  
  if (userType === 'email') {
    // user id validation for hex pattern of email login
    return userHexIdPattern.test(userID);
  } else {
    // user id validation for anonymous login
    return userIDPattern.test(userID);
  }
};
```
Server function
```
const validateUserID = (req, res, next) => {
  const {id} = req.body;
  const userIDPattern = /^[a-z0-9]{12}$/;

  if(id !== undefined && (typeof id !== 'string' || !userIDPattern.test(id))){
    //if id of type string is coming it'll be via the anonymous validation
    return res.status(NOT_ACCEPTABLE).json({
      message: 'Invalid login Id.'
    });
  }else{
    // if id is not there or the id is valid move ahead
    next();
  }
}
```
# ✔️ Check List (Check all the applicable boxes)

- [X] My code follows the code style of this project.
- [X] This PR does not contain plagiarized content.
- [X] The title and description of the PR is clear and explains the approach.

##  Note to reviewers
Yes regarding the userId type I've checked the id generated by the function `crypto.randomBytes(12).toString('hex');` was of 24 char and the regex as mentioned above, that's why needed to add extra regex check for the hex id pattern.

# 📷 Screenshots

NA

### CODE REVIEW

File: client/src/context/AuthContext.jsx [hunk 1]
```diff
@@ -22,11 +23,10 @@ const initializeAuthState = (defaultState) => {
 		if (!persistedState) {
 			return defaultState;
 		}
-
-		/**
-		 * TODO: Validate loginId to be a valid uuid using the uuid
-		 *       library from npm
-		 */
+		if (!validateUserID(persistedState.loginId, persistedState.loginType)) {
```
Comment (Dun-sin) [inline client/src/context/AuthContext.jsx:26]:
```suggestion

		if (!validateUserID(persistedState.loginId, persistedState.loginType)) {
```

File: client/src/lib/utils.js [hunk 2]
```diff
@@ -46,3 +46,16 @@ export const decrypt = (encryptedText) => {
...
+	const userHexIdPattern = /^[a-f0-9]{24}$/;
+	
+	if (userType === 'email') {
+		// user id validation for hex pattern of email login
+		return userHexIdPattern.test(userID);
+	} else {
+		// user id validation for anonymous login
+		return userIDPattern.test(userID);
+	}
+};
```
Comment (Dun-sin) [inline client/src/lib/utils.js:61]:
```suggestion
};

```

File: server/utils/helper.js [hunk 3]
```diff
@@ -18,7 +19,23 @@ function generateObjectId() {
...
+    //if id of type string is coming it'll be via the anonymous validation
+    return res.status(NOT_ACCEPTABLE).json({
+      message: 'Invalid login Id.'
+    });
+  }else{
+    // if id is not there or the id is valid move ahead
+    next();
+  }
+}
+
```
Comment (Dun-sin) [inline server/utils/helper.js:36]:
```suggestion
const validateUserID = (req, res, next) => {
  const { id } = req.body;
  const userIDPattern = /^[a-z0-9]{12}$/;

  // If id is required and not present or invalid, reject the request
  if (!id || typeof id !== 'string' || !userIDPattern.test(id)) {
    return res.status(406).json({
      message: 'Invalid login Id.'
    });
  }

  // If id is valid, proceed to the next middleware
  next();
}

```

**Review by Dun-sin — APPROVED** [review]
You did great @Gopal-001, thanks for contributing, I hope you will stick around and continue to contribute to this project.

**Comment by Gopal-001** [comment]
> You did great @Gopal-001, thanks for contributing, I hope you will stick around and continue to contribute to this project.

Thank you so much @Dun-sin ! It’s been a pleasure contributing, and I look forward to learning and adding more value to the project. 😊

**Pull request merged**